### PR TITLE
Add autograd tests for complex matrix norm nuclear and +/-2

### LINKED ISSUE #48842: Support autograd in `torch.svd` with complex inputs
## 🚀 Feature
`torch.svd` does not have autograd support for complex inputs.

## Motivation

While working on updating complex support for `torch.linalg.norm` and `torch.norm`, I noticed that the code paths that use `at::svd` underneath were throwing errors when given a complex input that requires grad, like the following:

```
>>> a = torch.randn(10, 10, dtype=torch.cfloat, requires_grad=True)
>>> r = torch.svd(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: svd does not support automatic differentiation for outputs with complex dtype.
```

## Pitch

Support autograd with complex inputs

cc @ezyang @anjali411 @dylanbespalko @mruberry

### PR DESCRIPTION
**Pull request by kurtamohler**
Also upgrades `linalg.norm`'s autograd and jit tests to `OpInfo`

Fixes #48842

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -1327,29 +1328,49 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
...
             res.backward()
 
-            x_real = x.clone().detach().abs().requires_grad_(True)
-            res_real = torch.linalg.norm(x_real, ord, keepdim=keepdim)
-            res_real.backward()
+            is_nuclear = isinstance(ord, str) and ord == 'nuc'
+            is_matrix_2 = isinstance(ord, Number) and abs(ord) == 2
 
-            msg = gen_error_message(x.size(), ord, keepdim)
+            # Matrix norms order +/-2 and Nuclear can be compared with SVD
```
Comment (anjali411) [inline test/test_linalg.py:1342]: @kurtamohler why don't we just use `gradcheck` here? We should also consider adding an `OpInfo` for `torch.linalg.norm` I think 

cc. @mruberry

Comment (kurtamohler) [reply test/test_linalg.py:1342]: For some reason I was under the impression that `gradcheck` only works for `torch.double`, but it does indeed work with `torch.cdouble` as well. However, it does not work for `torch.float` or `torch.cfloat`: https://github.com/pytorch/pytorch/blob/master/torch/autograd/gradcheck.py#L309

I assume it's acceptable not to include autograd tests for `torch.float` and `torch.cfloat`, so I'll update to use `gradcheck`. In that case, I can actually just remove `test_norm_complex_autograd` and add `torch.cdouble` to `test_autograd_and_jit`, which already uses `gradcheck`.

Comment (mruberry) [reply test/test_linalg.py:1342]: You're correct, @kurtamohler, that it's only intended to support double and cdouble, and we typically only test grad and gradgrad in those dtypes.

I think it is time to implement an OpInfo for torch.linalg.norm. That will automatically generate grad and gradgrad tests for all of its sample inputs. Ping me if you have questions about how to do this.

Comment (kurtamohler) [reply test/test_linalg.py]: But yes, `OpInfo` seems like a good idea. I'll look into that now

**Comment by kurtamohler** [comment]
I've replaced `linalg.norm`'s autograd and jit tests with an `OpInfo` entry. Should be ready for review

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1116,6 +1178,22 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
            sample_inputs_func=sample_inputs_fliplr_flipud,
            test_inplace_grad=False,
            supports_tensor_out=False),
+    OpInfo('linalg.norm',
+           op=torch.linalg.norm,
+           dtypes=floating_and_complex_types_and(torch.float16, torch.bfloat16),
+           test_inplace_grad=False,
+           test_complex_grad=True,
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1185]: `test_complex_grad=True` by default so you can remove this if you like

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -1116,6 +1178,22 @@ def sample_inputs_fliplr_flipud(op_info, device, dtype, requires_grad):
...
+           supports_tensor_out=True,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
+           sample_inputs_func=sample_inputs_linalg_norm,
+           aten_name='linalg_norm',
+           skips=(
+               # TODO: remove this once `pow` is implemented for float16
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1191]: it would be awesome if you could file an issue and link it here

**Review by anjali411 — APPROVED** [review]
left some minor comments, but looks good otherwise :D thank you and let me know once you link the issue for pow inline.

**Comment by kurtamohler** [comment]
Thanks @anjali411, I've made those changes. I made an issue for `pow` and for `max`.

CI was showing a failing mypy test related to my changes. I'm not sure yet if it's fixed it or not, we'll see.

**Comment by anjali411** [comment]
Hey @kurtamohler could you rebase? I think there might be merge conflicts due to #50667 which landed earlier today

**Comment by kurtamohler** [comment]
@anjali411, done. If there were conflicts, git handled them automatically for me

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:1256]: #50789

**Pull request closed without merging**